### PR TITLE
Unicode

### PR DESCRIPTION
**Pull request by ocefpaf**
@lsetiawan this is not ready for review yet. I am not sure if it will solve our problems.

BTW, do we have the bug as a test case? B/c the tests were passing before, right?

### CODE REVIEW

File: test/test_wof_1_1.py [hunk 1]
```diff
@@ -47,7 +47,7 @@ def response_to_StringIO(self, response, root_name):
         """
         Converts WOF method response to StringIO object.
         """
-        response_output = io.BytesIO()
+        response_output = io.StringIO()
```
Comment (ocefpaf) [inline test/test_wof_1_1.py]: It seems that this test expected `StringIO`, we can revisit once we move to Python 3.

File: wof/WaterML_1_1.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 # Generated Thu Oct 01 11:55:06 2015 by generateDS.py version 2.17a.
 #
 # Command line options:
-#   ('-o', 'wof\\WaterML_1_1.py')
+#   (u'-o', u'wof\\WaterML_1_1.py')
```
Comment (ocefpaf) [inline wof/WaterML_1_1.py]: It is a gigantic diff and GitHub even refuses to render :smile: 
But there are some noise from automate replaces. Just ignore that.

File: wof/WaterML_1_1.py [hunk 3]
```diff
@@ -18,14 +18,14 @@
 #   WOFpy
 #
 
-from __future__ import (absolute_import, division, print_function)
+from __future__ import (absolute_import, division, print_function, unicode_literals)
```
Comment (ocefpaf) [inline wof/WaterML_1_1.py]: I am not a fan of forcing `unicode_literals`. But if it works...

File: wof/WaterML_1_1.py [hunk 4]
```diff
@@ -18,14 +18,14 @@
...
 
 import sys
-import re as re_
+import re
 import base64
-import datetime as datetime_
-import warnings as warnings_
-from lxml import etree as etree_
```
Comment (ocefpaf) [inline wof/WaterML_1_1.py]: Sorry for the noise. I just never got why the original imports were renamed to something strange.
Maybe someone trying to make the code more obscure to his job :laughing:

**Comment by lsetiawan** [comment]
> BTW, do we have the bug as a test case? B/c the tests were passing before, right?

I did make a test before #156 ... Not sure how robust it is though.

File: wof/WaterML_1_1.py [hunk 5]
```diff
@@ -402,10 +402,11 @@ def quote_xml(inStr):
         pos = mo.end()
     s3 = s1[pos:]
     s2 += quote_xml_aux(s3)
-    return s2.decode(ExternalEncoding)
```
Comment (ocefpaf) [inline wof/WaterML_1_1.py]: This is what we should avoid, the encode/decode whack-a-mole :smile: 

[The principle is:](https://docs.python.org/2/howto/pyporting.html)

```
Decide which of your APIs take text and which take binary data
Make sure that your code that works with text also works with unicode and code for binary data works with bytes in Python 2 (see the table above for what methods you cannot use for each type)
Mark all binary literals with a b prefix, textual literals with a u prefix
Decode binary data to text as soon as possible, encode text as binary data as late as possible
Open files using io.open() and make sure to specify the b mode when appropriate
Be careful when indexing into binary data
```

File: wof/WaterML_1_1.py [hunk 6]
```diff
@@ -402,10 +402,11 @@ def quote_xml(inStr):
...
+    return s2
 
 
 def quote_xml_aux(inStr):
+    inStr = inStr.decode('utf-8')
```
Comment (ocefpaf) [inline wof/WaterML_1_1.py]: If I got the code correctly `inStr` comes from the data read, so here we must decode, work with strings, and encode at the end if necessary.

**Comment by ocefpaf** [comment]
> I did make a test before #156 ... Not sure how robust it is though.

I take a weak test over nothing any day :smile: 

We can try that and @sreeder's bug report.

**Comment by lsetiawan** [comment]
@ocefpaf OMG IT'S PASSING!! 😮  YOU ARE AWESOME!

**Comment by ocefpaf** [comment]
> @ocefpaf OMG IT'S PASSING!! 😮  YOU ARE AWESOME!

You mean passing @sreeder case? I don't really think our tests here caught the issue in the first place :smile:

**Comment by lsetiawan** [comment]
> You mean passing @sreeder case? I don't really think our tests here caught the issue in the first place 

Idk. Lemme test locally.

**Comment by ocefpaf** [comment]
GitHub cannot even handle the diff :rofl: 

screenshot from 2017-09-01 16-54-52

File: wof/WaterML_1_1.py [hunk 7]
```diff
@@ -4605,9 +4605,9 @@ def validate_definitionType(self, value):
         # Validate type definitionType, a restriction on xsi:string.
         if value is not None and Validate_simpletypes_:
             if len(value) > 255:
-                warnings.warn(u'Value "%(value)s" does not match xsd maxLength restriction on definitionType' % {"value" : value.encode("utf-8")} )
+                warnings.warn(u'Value "%(value)s" does not match xsd maxLength restriction on definitionType' % {"value" : value} )
             if len(value) < 0:
-                warnings.warn(u'Value "%(value)s" does not match xsd minLength restriction on definitionType' % {"value" : value.encode("utf-8")} )
+                warnings.warn(u'Value "%(value)s" does not match xsd minLength restriction on definitionType' % {"value" : value} )
```
Comment (ocefpaf) [inline wof/WaterML_1_1.py]: If this works we have a single `.decode` point at

```
def quote_xml_aux(inStr):
    inStr = inStr.decode('utf-8')
    s1 = inStr.replace('&', '&amp;')
    s1 = s1.replace('<', '&lt;')
    s1 = s1.replace('>', '&gt;')
    return s1
```

and no re-encoding. It should be all unicode/str (py2/py3).

We may still write it as bytes in the future.

**Comment by lsetiawan** [comment]
@ocefpaf I tested similar request like @sreeder test case

`.../postgresqlodm2timeseries/rest/1_1/GetValues?location=postgresqlodm2timeseries:desk_sensors&variable=postgresqlodm2timeseries:Decagon_CTD-10_Temp`

From her description:
> The character is the trademark symbol located in the organization definition field. It is connected to a postgres database.

I added: `Decagon Devices™` to Organization Description.

Now I get the error below, running the code with this PR's branch.

```xml
<ns0:Fault xmlns:ns0="http://schemas.xmlsoap.org/soap/envelope/">
<faultcode>soap11env:Server</faultcode>
<faultstring>'unicode' does not have the buffer interface</faultstring>
<faultactor/>
</ns0:Fault>
```

**Comment by lsetiawan** [comment]
Here's the full traceback:

```python
ERROR:spyne.application.server:'unicode' does not have the buffer interface
Traceback (most recent call last):
  File "/home/ubuntu/miniconda/envs/wofpydev/lib/python2.7/site-packages/spyne/application.py", line 151, in process_request
    ctx.out_object = self.call_wrapper(ctx)
  File "/home/ubuntu/miniconda/envs/wofpydev/lib/python2.7/site-packages/spyne/application.py", line 235, in call_wrapper
    retval = ctx.descriptor.service_class.call_wrapper(ctx)
  File "/home/ubuntu/miniconda/envs/wofpydev/lib/python2.7/site-packages/spyne/service.py", line 209, in call_wrapper
    return ctx.function(ctx, *args)
  File "/home/ubuntu/wofpy-dev/wof/apps/spyned_1_1.py", line 208, in GetValues
    authToken
  File "/home/ubuntu/wofpy-dev/wof/apps/spyned_1_1.py", line 184, in GetValuesObject
    namespacedef_=NSDEF)
  File "/home/ubuntu/wofpy-dev/wof/WaterML_1_1.py", line 3472, in export
    outfile.write(u'<%s%s%s' % (namespace_, name_, namespacedef_ and ' ' + namespacedef_ or '', ))
TypeError: 'unicode' does not have the buffer interface
```

**Comment by ocefpaf** [comment]
Thanks! That is good. We need to add that as a test here in order to make progress.

**Comment by lsetiawan** [comment]
@ocefpaf The current test is actually capturing the encoding/decoding error. Your change https://github.com/ODM2/WOFpy/pull/188#discussion-diff-136649344R50 makes the test different from the actual application https://github.com/ODM2/WOFpy/blob/master/wof/apps/spyned_1_1.py#L38. This is why test is passing while application is failing. When I changed the test back to use `io.BytesIO()`. I get test failures of 

```python
...
>       outfile.write(u'<%s%s%s' % (namespace_, name_, namespacedef_ and ' ' + namespacedef_ or '', ))
E       TypeError: 'unicode' does not have the buffer interface

wof/WaterML_1_1.py:3472: TypeError
```

**Comment by lsetiawan** [comment]
When I changed the `BytesIO()` to `StringIO()` on the actual application, this is my traceback error:

```python
  File "/home/ubuntu/wofpy-dev/wof/WaterML_1_1.py", line 408, in quote_xml_aux
    inStr = inStr.decode('utf-8')
  File "/home/ubuntu/miniconda/envs/wofpydev/lib/python2.7/encodings/utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2122' in position 18: ordinal not in range(128)
```

**Comment by lsetiawan** [comment]
To follow up https://github.com/ODM2/WOFpy/pull/188#issuecomment-326686420.

I know I'm not supposed to do unicode whack-a-mole, but when I encode then decode `inStr` The problem goes away. `inStr = inStr.encode('utf-8').decode('utf-8')` This is also changing the `BytesIO` to `StringIO`.

**Comment by lsetiawan** [comment]
Looking into things deeper, it appears that the type of string for `inStr` is `unicode`.

**Comment by ocefpaf** [comment]
> I know I'm not supposed to do unicode whack-a-mole, but when I encode then decode `inStr` The problem goes away. `inStr = inStr.encode('utf-8').decode('utf-8')` This is also changing the `BytesIO` to `StringIO`.

Changing only that is not a problem b/c `inStr` is the data entrypoint and `BytesIO` or `StringIO` is the endpoint. I want to avoid are internal encoding/decoding. However, to change that we need to be sure that is expected at those points.

**Comment by lsetiawan** [comment]
What's the status of this PR?

**Comment by ocefpaf** [comment]
> What's the status of this PR?

Unresolved. Here is my take on this:

- this PR cannot address the issue until we "beef up" the tests to actually reproduce the errors; the green status here means nothing;
- the revert to StringIO in the tests were due to the "nature" of the tests, which does not give us much bearing actually as I don't really know what are the excepted outputs there: bytes or something else?

If your fix worked for @sreeder setup go with that and close this one until we can answer those questions and lay a strategy to move forward.

**Comment by emiliom** [comment]
> this PR cannot address the issue until we "beef up" the tests to actually reproduce the errors; the green status here means nothing;

What's the specific github issue or bug originally addressed by this PR? I've lost track. (I realize there is a broader unicode handling problem in wofpy, but it'd still be helpful to know what specific error report generated this PR, despite the whack-a-mole situation)

> If your fix worked for @sreeder setup go with that and close this one until we can answer those questions and lay a strategy to move forward.

FYI, Stephanie is starting maternity leave this week, probably today. I don't know if there's anyone in Jeff's team who will be able to continue to do wofpy implementation testing and engagement with us during Stephanie's absence.

**Comment by lsetiawan** [comment]
> What's the specific github issue or bug originally addressed by this PR?

https://github.com/ODM2/WOFpy/issues/148#issuecomment-326420842

**Comment by sreeder** [comment]
@lsetiawan @emiliom  I am still in today, I will be out tomorrow though. So if you guys would like me to test out this issue let me know!

**Comment by lsetiawan** [comment]
@sreeder Go ahead and try out this branch and see if it works for you. Thanks.

**Comment by sreeder** [comment]
@lsetiawan  this is going to sound really novice, but how do I check out this branch since it is not on the ODM2 wofpy repo? :-)

**Comment by ocefpaf** [comment]
@sreeder this should work:

```shell
git clone https://github.com/ocefpaf/WOFpy.git
cd WOFpy
git checkout unicode
```

**Comment by sreeder** [comment]
@ocefpaf @lsetiawan @emiliom  
I am still getting an error: but it is a different one. 
screen shot 2017-09-18 at 2 31 27 pm

**Comment by ocefpaf** [comment]
> I am still getting an error: but it is a different one.

Thanks for testing this @sreeder. Unless we can get that application as a test in this repo will be driving blind here. @lsetiawan can we have a call sometime to see if we can set that test up?

**Comment by sreeder** [comment]
@ocefpaf  would it help if I gave you the exact text that is failing in my database?

**Comment by ocefpaf** [comment]
> @ocefpaf  would it help if I gave you the exact text that is failing from my database?

Thanks @sreeder. The main blocker, at least to me, is to set a database on Travis-CI for the testing. I am looking into it.

**Comment by lsetiawan** [comment]
@ocefpaf Can we use the testing method like ODM2PythonAPI. That repo spins up a database within Travis.

**Comment by lsetiawan** [comment]
Like I said on https://github.com/ODM2/WOFpy/pull/188#issuecomment-326689885. The current test mechanic is different than the actual application currently.

**Comment by ocefpaf** [comment]
> @ocefpaf Can we use the testing method like ODM2PythonAPI? That repo spins up a database within Travis.

That is exactly what I was going to try :smile:

**Comment by lsetiawan** [comment]
I am going to merge this PR for now and further change things as I mention in https://github.com/ODM2/WOFpy/pull/188#issuecomment-326686420

**Comment by miguelcleon** [comment]
@lsetiawan should I try installing this update?

**Comment by ocefpaf** [comment]
> @lsetiawan should I try installing this update?

@miguelcleon please let us know the results if you do. This part of the code needs more testing and supporting `unicode` for Python 2 is still a complicated task.

**Comment by miguelcleon** [comment]
Ok now I got a different error from: 
http://odm2admin.cuahsi.org/wofpy/odm2timeseries/rest/1_1/GetSiteInfo?site=odm2timeseries:Rio%20Icacos%20Trib-IO 

```
<ns0:Fault xmlns:ns0="http://schemas.xmlsoap.org/soap/envelope/">
<faultcode>soap11env:Server</faultcode>
<faultstring>'unicode' does not have the buffer interface</faultstring>
<faultactor/>
</ns0:Fault>
```

**Comment by lsetiawan** [comment]
@miguelcleon Go ahead and try my latest change. Thanks!
